tests/TestCase/Http/Middleware/RateLimitMiddlewareTest.php

```diff
@@ -0,0 +1,347 @@
+        ]);
+
+        $response = $middleware->process($request, $this->handler);
+        $this->assertFalse($response->hasHeader('X-RateLimit-Limit'));
+        $this->assertFalse($response->hasHeader('X-RateLimit-Remaining'));
+        $this->assertFalse($response->hasHeader('X-RateLimit-Reset'));
+    }
+
+    /**
+     * Test user-based identification
+     *
+     * @return void
+     */
+    public function testUserIdentification(): void
+    {
+        $middleware = new RateLimitMiddleware([
+            'limit' => 1,
+            'window' => 60,
+            'identifier' => 'user',
+            'cache' => 'rate_limit_test',
+        ]);
+
+        $identity = new stdClass();
+        $identity->id = '123';
+
+        $request = (new ServerRequest([
+            'environment' => ['REMOTE_ADDR' => '127.0.0.1'],
+        ]))->withAttribute('identity', $identity);
+
+        $response = $middleware->process($request, $this->handler);
+        $this->assertEquals('0', $response->getHeaderLine('X-RateLimit-Remaining'));
+
+        // Second request should fail
+        $this->expectException(TooManyRequestsException::class);
+        $middleware->process($request, $this->handler);
+    }
+
+    /**
+     * Test route-based identification
+     *
+     * @return void
+     */
+    public function testRouteIdentification(): void
+    {
+        $middleware = new RateLimitMiddleware([
+            'limit' => 1,
+            'window' => 60,
+            'identifier' => 'route',
+            'cache' => 'rate_limit_test',
+        ]);
+
+        $request = (new ServerRequest([
+            'environment' => ['REMOTE_ADDR' => '127.0.0.1'],
+        ]))->withAttribute('params', [
+            'controller' => 'Users',
+            'action' => 'login',
+        ]);
+
+        $response = $middleware->process($request, $this->handler);
+        $this->assertEquals('0', $response->getHeaderLine('X-RateLimit-Remaining'));
+
+        // Different route should work
+        $request2 = (new ServerRequest([
+            'environment' => ['REMOTE_ADDR' => '127.0.0.1'],
+        ]))->withAttribute('params', [
+            'controller' => 'Users',
+            'action' => 'logout',
+        ]);
+
+        $middleware->process($request2, $this->handler);
+
+        // Same route should fail
+        $this->expectException(TooManyRequestsException::class);
+        $middleware->process($request, $this->handler);
+    }
+
+    /**
+     * Test proxy headers for IP detection
+     *
+     * @return void
+     */
+    public function testProxyHeaders(): void
+    {
+        $middleware = new RateLimitMiddleware([
+            'limit' => 1,
+            'window' => 60,
+            'cache' => 'rate_limit_test',
+        ]);
+
+        // Test Cloudflare header
+        $request = new ServerRequest([
+            'environment' => [
+                'REMOTE_ADDR' => '127.0.0.1',
+                'HTTP_CF_CONNECTING_IP' => '192.168.1.100',
```

Comment:
Why are we including paid product/service integrations? A more generic solution could be to allow developers to define which header their forwarding IP is defined in, and default to `x-forwarded-for`